### DISCUSSION TITLE
pupil_rays in PupilSpec

### DISCUSSION THREAD

**Discussion by dibyendumajumdar** [mjhoptics/ray-optics#80] [Q&A] [answered]
Hi @mjhoptics 

Trying to understand pupil_rays. I have been messing around with code like this:

```
layout = LensLayout(opm)
rays = layout.create_ray_entities(None, 0.0)
for ray in rays:
    ray.update_shape(None)
```

I noticed that one of the fields failed to complete the ray trace, which didn't make sense to me, as I imagined that +X,-X,+Y,-Y are all equivalent in terms of the ray tracing.

**Comment by mjhoptics** [comment]
The pupil rays will (normally) only be equivalent for an axial field point. Once you go off-axis, the pupil is typically vignetted. The upper and lower rays (+Y and -Y) will be unequally limited; the +/-X rays will be less restricted.

The code associated with Lens Layout has a bunch of stuff related to drawing the rays. If you just want the data, you can use the functions in the `raytr.trace` package, e.g.
```
rays = trace.trace_boundary_rays(opm)
```

**Reply by dibyendumajumdar** [reply]
> The pupil rays will (normally) only be equivalent for an axial field point. Once you go off-axis, the pupil is typically vignetted. The upper and lower rays (+Y and -Y) will be unequally limited; the +/-X rays will be less restricted.

Apart from the fact that Y and X are on different axes, why would there be a difference? As it lens is circular, I expected that if X+ works, then it implies X-, Y+, Y- should also work. I am assuming a circular image plane that covers the whole field.
 
> The code associated with Lens Layout has a bunch of stuff related to drawing the rays. If you just want the data, you can use the functions in the `raytr.trace` package, e.g.
> 
> ```
> rays = trace.trace_boundary_rays(opm)
> ```

Yes, thank you.  I commented out the gui stuff in LensLayout, can't remember now why I wanted to do it from there.

**Reply by mjhoptics** [reply]
Once a field point moves off axis (for example in the Y direction), the system no longer appears symmetric to that object point; it is up/down asymmetric. Depending on the design of the lens, the Y vignetting can be asymmetric or symmetric but it will typically not pass the full fan of rays. The X rays are less affected when going off axis so little or no vignetting is typically observed.

The Double Gauss form shows this behavior. Note that the upper and lower pupil rays at full field don't fill the aperture stop; they are clipped by the edges of the first and last lenses:
<img width="740" alt="Screen Shot 2022-02-20 at 3 57 30 PM" src="https://user-images.githubusercontent.com/33294031/154870288-004a45e7-460e-4b59-8c3e-d19492e62477.png">

The transverse ray aberration curves show the Y ray fans on the left get vignetted as you move off axis; the X fans on the right are not limited much or at all.
<img width="540" alt="Screen Shot 2022-02-20 at 4 01 35 PM" src="https://user-images.githubusercontent.com/33294031/154870378-030dc524-4167-41c7-ba3c-00639327ae13.png">

**Reply by dibyendumajumdar** [reply]
I think I am misunderstanding the meaning of the X and Y labels and the +1,-1 on the pupil rays.

**Reply by mjhoptics** [reply, marked as answer]
Let me try to describe what is being plotted in the graph above.

Each horizontal pair of graphs show data from a single point in the field of view; the bottom row is on-axis, the middle row is at 0.7 full field and the top row is the full/edge of field.

The horizontal axis of the plots is ray position in the entrance pupil. The left side plot shows ray position along the Y (vertical) axis of the entrance pupil; the right side plot shows ray position along the X (horizontal) axis of the pupil.

The vertical axis of the plots shows the ray aberration in the image plane. So the plots are showing ray position in the image plane as a function of position in the entrance pupil. The left side plots show the difference in Y ray coordinate in the image plane between the chief (central) ray at the central (green) wavelength and a ray traced through the Y axis of the pupil. Similarly, the right side plots show the difference in X ray coordinate in the image plane between the chief ray at the central wavelength and a ray traced through the X axis of the pupil. The data so produced has characteristic shapes for different aberrations. For instance, on axis the dominant aberration is spherical aberration with some residual chromatic aberration. 

For the axial field point (bottom row), the aberration produced by tracing rays through the X and Y axes of the pupil are identical due to the rotational symmetry of the situation. The other field points are displaced along the Y axis of the object plane, i.e. off of the symmetry axis of the lens. Now the ray fans through the X and Y axes of the entrance pupil are different and show different aberration properties. Note that for a Y field point, the system is still left/right symmetric (+X == -X) and so the aberration plots of the right hand column show left/right symmetry. The left hand column is asymmetric in Y, i.e. +Y != -Y, and so the aberration plots off axis are asymmetric.

Note: The plot I included above exposes a bug that I fixed last night. It should look like this:
<img width="538" alt="Screen Shot 2022-02-21 at 11 52 27 AM" src="https://user-images.githubusercontent.com/33294031/155018491-1b9ec1dc-95bd-4a08-b48e-8f225f3f7b3e.png">

**Reply by dibyendumajumdar** [reply]
Hi @mjhoptics 

Is it possible to visualize the ray fans?

**Reply by dibyendumajumdar** [reply]
I tried visualizing the rays in 3D in Zemax. I think I now get what you were saying above. Here are couple of images - the lens has been hidden.

The rays on y axis (viewed from side):
![yaxis](https://github.com/user-attachments/assets/e4c19635-f50e-40a3-91c4-080322e5401b)

The rays on x axis (view from top):
![xaxis](https://github.com/user-attachments/assets/6a2b1a70-ff88-471a-b718-95cf32e48293)
